hack/test-go.sh

```diff
@@ -109,87 +109,99 @@ if [[ ${#testcases[@]} -eq 0 ]]; then
-      fi
-      count=$((count + 1))
-    done 2>&1
-    kube::log::status "${pass} / ${count} passed"
-  done
-  if [[ ${fails} -gt 0 ]]; then
-    exit 1
-  else
+
+  # If we're not collecting coverage, run all requested tests with one 'go test'
+  # command, which is much faster.
+  if [[ ! ${KUBE_COVER} =~ ^[yY]$ ]]; then
+    kube::log::status "Running unit tests without code coverage"
+    go test "${goflags[@]:+${goflags[@]}}" \
+      ${KUBE_RACE} ${KUBE_TIMEOUT} "${@+${@/#/${KUBE_GO_PACKAGE}/}}"
     exit 0
   fi
-fi
 
-# If we're not collecting coverage, run all requested tests with one 'go test'
-# command, which is much faster.
-if [[ ! ${KUBE_COVER} =~ ^[yY]$ ]]; then
-  kube::log::status "Running unit tests without code coverage"
-  go test "${goflags[@]:+${goflags[@]}}" \
-    ${KUBE_RACE} ${KUBE_TIMEOUT} "${@+${@/#/${KUBE_GO_PACKAGE}/}}"
-  exit 0
-fi
+  # Create coverage report directories.
+  cover_report_dir="/tmp/k8s_coverage/$(kube::util::sortable_date)"
+  cover_profile="coverage.out"  # Name for each individual coverage profile
+  kube::log::status "Saving coverage output in '${cover_report_dir}'"
+  mkdir -p "${@+${@/#/${cover_report_dir}/}}"
+
+  # Run all specified tests, collecting coverage results. Go currently doesn't
+  # support collecting coverage across multiple packages at once, so we must issue
+  # separate 'go test' commands for each package and then combine at the end.
+  # To speed things up considerably, we can at least use xargs -P to run multiple
+  # 'go test' commands at once.
+  printf "%s\n" "${@}" | xargs -I{} -n1 -P${KUBE_COVERPROCS} \
+      go test "${goflags[@]:+${goflags[@]}}" \
+          ${KUBE_RACE} \
+          ${KUBE_TIMEOUT} \
+          -cover -covermode="${KUBE_COVERMODE}" \
+          -coverprofile="${cover_report_dir}/{}/${cover_profile}" \
+          "${cover_params[@]+${cover_params[@]}}" \
+          "${KUBE_GO_PACKAGE}/{}"
+
+  combined_cover_profile="${cover_report_dir}/combined-coverage.out"
+  {
+    # The combined coverage profile needs to start with a line indicating which
+    # coverage mode was used (set, count, or atomic). This line is included in
+    # each of the coverage profiles generated when running 'go test -cover', but
+    # we strip these lines out when combining so that there's only one.
+    echo "mode: ${KUBE_COVERMODE}"
+
+    # Include all coverage reach data in the combined profile, but exclude the
+    # 'mode' lines, as there should be only one.
+    for x in `find "${cover_report_dir}" -name "${cover_profile}"`; do
+      cat $x | grep -h -v "^mode:" || true
+    done
+  } >"${combined_cover_profile}"
+
+  coverage_html_file="${cover_report_dir}/combined-coverage.html"
+  go tool cover -html="${combined_cover_profile}" -o="${coverage_html_file}"
+  kube::log::status "Combined coverage report: ${coverage_html_file}"
+
+  if [[ -x "${KUBE_GOVERALLS_BIN}" ]]; then
+    ${KUBE_GOVERALLS_BIN} -coverprofile="${combined_cover_profile}" || true
```

Comment:
This is where a coverage report is sent to Coveralls - so if you call `runTests()` multiple times, it's going to report coverage there multiple times.

We probably only want to be reporting coverage once.